### PR TITLE
Cleanup path handling

### PR DESCRIPTION
**Pull request by KarlStraussberger**
Simplify map access

### CODE REVIEW

File: src/util/tools.h [hunk 1]
```diff
@@ -87,17 +87,12 @@ std::string reduceString(std::string str, char ch);
...
 
-/// \brief Get last write time of the specified file or path, if it does not exist it will throw an exception
-/// \param path file or directory to be checked.
-/// \return last modification time of the path or directory
-time_t getLastWriteTime(const fs::path& path);
-
 /// \brief Checks if the given file is a regular file (imitate same behaviour as std::filesystem::is_regular_file)
-bool isRegularFile(const fs::path& path);
 bool isRegularFile(const fs::path& path, std::error_code& ec) noexcept;
+bool isRegularFile(const fs::directory_entry& dirEnt, std::error_code& ec) noexcept;
```
Comment (neheb) [inline src/util/tools.h:92]: Why not replace this with a template?

File: src/config/config_setup.cc [hunk 2]
```diff
@@ -184,7 +184,9 @@ bool ConfigPathSetup::checkAgentPath(std::string& optValue)
 {
     std::string tmp_path;
     if (fs::path(optValue).is_absolute()) {
-        if (!isRegularFile(optValue) && !fs::is_symlink(optValue)) {
+        std::error_code ec;
+        fs::directory_entry dirEnt(optValue, ec);
+        if (!isRegularFile(dirEnt, ec) && !dirEnt.is_symlink(ec)) {
```
Comment (neheb) [inline src/config/config_setup.cc:189]: I would rather use fs::is_symlink(dirEnt, ec) here and elsewhere.

It's more generic and helps reduce changes when refactoring.

File: src/content/content_manager.cc [hunk 3]
```diff
@@ -315,8 +323,9 @@ void ContentManager::shutdown()
             log_debug("AutoScanDir {}", i);
             std::shared_ptr<AutoscanDirectory> dir = autoscan_inotify->get(i);
             if (dir != nullptr) {
-                if (fs::is_directory(dir->getLocation())) {
-                    auto t = getLastWriteTime(dir->getLocation());
+                auto dirEnt = fs::directory_entry(dir->getLocation());
+                if (dirEnt.is_directory()) {
+                    auto t = to_time_t(dirEnt.last_write_time());
                     dir->setCurrentLMT(dir->getLocation(), t);
```
Comment (neheb) [inline src/content/content_manager.cc:329]: setCurrentLMT should probably be taking std::chrono::seconds or something.

File: src/util/tools.cc [hunk 4]
```diff
@@ -230,23 +217,23 @@ bool isRegularFile(const fs::path& path, std::error_code& ec) noexcept
...
-    int ret = stat(path.c_str(), &statbuf);
+    int ret = stat(dirEnt.path().c_str(), &statbuf);
     if (ret != 0) {
         throw_std_runtime_error("{}: {}", std::strerror(errno), path.c_str());
     }
 
     return statbuf.st_size;
 #else
-    return fs::file_size(path);
+    return dirEnt.file_size();
```
Comment (neheb) [inline src/util/tools.cc:236]: What's the difference here?

File: src/util/tools.h (hunk 1 above)
Comment (KarlStraussberger) [reply src/util/tools.h:92]: Because the code is not generic.

File: src/config/config_setup.cc (hunk 2 above)
Comment (KarlStraussberger) [reply src/config/config_setup.cc:189]: The point here is the sequence of `stats`. directory_entry encapsulates this.

File: src/content/content_manager.cc (hunk 3 above)
Comment (KarlStraussberger) [reply src/content/content_manager.cc:329]: Good point. The focus was on filesystem operations. I put it on my todo list.

File: src/util/tools.cc (hunk 4 above)
Comment (KarlStraussberger) [reply src/util/tools.cc:236]: The scanning loop already has `directory_entry` and the constructor does the `stat`s. By using the method we avoid doing it all over again. The static method only takes `path`. That's the core of the cleanup: use `directory_entry` if already created.

**Comment by whyman** [comment]
LGTM! Needs a rebase though, otherwise happy for it to go in. Looks good.

**Review by whyman — APPROVED** [review]

**Comment by whyman** [comment]
@KarlStraussberger it doesnt like the rebase, are you happy to sort or shall we squash instead?

**Comment by KarlStraussberger** [comment]
> @KarlStraussberger it doesnt like the rebase, are you happy to sort or shall we squash instead?

My PR got corrupted while on a rather slow remote connection, so rebasing gets some trouble.

**Pull request merged**